src/coreclr/nativeaot/System.Private.CoreLib/src/System/RuntimeExceptionHelpers.cs

```diff
@@ -235,72 +225,41 @@ internal static void FailFast(string message, Exception? exception, RhFailFastRe
-                    // Try to save the exception stack trace in a buffer on the stack.  If the exception is too large, we'll truncate it.
-                    const int MaxStack = 2048;
-                    Span<byte> exceptionStack = stackalloc byte[MaxStack];
-
-                    // Ignore output, as this is best-effort
-                    _ = Utf8.FromUtf16(outputMessage, exceptionStack, out _, out int length);
-                    // Fill the rest of the buffer with nulls
-                    if (length < MaxStack)
-                        exceptionStack.Slice(length).Clear();
-
-                    unsafe
-                    {
-                        byte* stackExceptionRecord = stackalloc byte[sizeof(CrashDumpRecord)];
-                        CrashDumpRecord* pExceptionRecord = (CrashDumpRecord*)stackExceptionRecord;
-                        var cookieSpan = new Span<byte>(pExceptionRecord->Cookie, CrashDumpRecord.CookieSize);
-                        // Random 10 bytes to identify the record
-                        ((ReadOnlySpan<byte>)new byte[] { 0x1c, 0x73, 0xd0, 0x2d, 0xda, 0x6b, 0x4c, 0xef, 0xbf, 0xa1 }).CopyTo(cookieSpan);
-                        "NETRUNTIME"u8.CopyTo(cookieSpan.Slice(10));
-                        pExceptionRecord->Type = 1;
-                        pExceptionRecord->Data = (void*)exceptionStack.GetPinnableReference();
-                        pExceptionRecord->Length = length;
-                    }
+                    crashInfo.WriteException(exception);
                 }
             }
 
-#if TARGET_WINDOWS
-            uint errorCode = 0x80004005; // E_FAIL
-            // To help enable testing to bucket the failures we choose one of the following as errorCode:
-            // * hashcode of EETypePtr if it is an unhandled managed exception
-            // * HRESULT, if available
-            // * RhFailFastReason, if it is one of the known reasons
-            if (exception != null)
-            {
-                if (reason == RhFailFastReason.UnhandledException)
-                    errorCode = (uint)(exception.GetEETypePtr().GetHashCode());
-                else if (exception.HResult != 0)
-                    errorCode = (uint)exception.HResult;
-            }
-            else if (reason != RhFailFastReason.Unknown)
+            crashInfo.Close();
+
+            // Try to map the failure into a HRESULT that makes sense
+            int errorCode = exception != null ? exception.HResult : reason switch
             {
-                errorCode = (uint)reason + 0x1000; // Add something to avoid common low level exit codes
-            }
+                RhFailFastReason.EnvironmentFailFast => HResults.COR_E_APPLICATION,
+                RhFailFastReason.InternalError or
+                RhFailFastReason.ClassLibDidNotTranslateExceptionID => HResults.COR_E_EXECUTIONENGINE,
+                RhFailFastReason.UnhandledException or
+                RhFailFastReason.UnhandledExceptionFromPInvoke or
```

Comment:
We do extract the hresult from the exception so that path should happen unless someone passes one of these RhFailFastReasons with an exception object.

Reply:
I'm not sure what the best HRESULT should be in this case.  It will be used as part of the bucketing.

Reply:
We should always get an exception object to get HRESULT from for `UnhandledException` and `UnhandledExceptionFromPInvoke`, so this should be unreachable. I think it would be fine to just use `HResults.COR_E_EXECUTIONENGINE` for all cases, except `EnvironmentFailFast`. All cases except `EnvironmentFailFast` mean that something went really wrong.